### PR TITLE
embeddings: featureflagged search against weaviate

### PR DESCRIPTION
**Pull request by keegancsmith**
This adds a feature flagged codepath to search against weaviate instead of our hand rolled embeddings searcher. This code is intended to be temporary while we validate weaviate. We have written a separate manual migration which takes data from our cloud buckets and inserts it into weaviate. So this avoids features like weaviate doing embeddings calls or keyword search. We just want to validate scale.

Test Plan: a bit clunky at the moment, but fine for our experimental phase. I mutated such that we just execute a search against the weaviate in prod. We create our test query by fetching an example vector from weaviate and double checking that its file appears in the search results.

```shellsession
$ kubectl port-forward -n prod weaviate-0  8080:8080
$ curl --request GET --url 'http://localhost:8080/v1/objects?class=Text_13482&include=vector&limit=1' | jq '.objects[0]' > Text_13482.json
$ cat Text_13482.json | jq .vector > enterprise/cmd/embeddings/shared/query.json
$ WEAVIATE_URL=http://localhost:8080 go run ./enterprise/cmd/embeddings
```

```diff
diff --git a/enterprise/cmd/embeddings/shared/main.go b/enterprise/cmd/embeddings/shared/main.go
index b152872280..c59784064f 100644
--- a/enterprise/cmd/embeddings/shared/main.go
+++ b/enterprise/cmd/embeddings/shared/main.go
@@ -8,6 +8,8 @@ import (
 	"net/http"
 	"time"
 
+	_ "embed"
+
 	"github.com/sourcegraph/log"
 
 	"github.com/sourcegraph/sourcegraph/cmd/frontend/globals"
@@ -39,6 +41,39 @@ import (
 
 const addr = ":9991"
 
+//go:embed query.json
+var queryJSON []byte
+
+func Main2(ctx context.Context, observationCtx *observation.Context, ready service.ReadyFunc, config *Config) error {
+	logger := observationCtx.Logger
+	getQueryEmbedding := func(ctx context.Context, query string) ([]float32, error) {
+		fmt.Println("query", query)
+		var vector []float32
+		err := json.Unmarshal(queryJSON, &vector)
+		return vector, err
+	}
+	readFile := func(ctx context.Context, repoName api.RepoName, revision api.CommitID, fileName string) ([]byte, error) {
+		fmt.Println("readFile", repoName, revision, fileName)
+		return nil, nil
+	}
+	w := newWeaviateClient(
+		logger,
+		readFile,
+		getQueryEmbedding,
+		config.WeaviateURL,
+	)
+
+	params := embeddings.EmbeddingsSearchParameters{
+		RepoID:           13482, // hashicorp/go-multierror repo id
+		Query:            "foo",
+		CodeResultsCount: 8,
+		TextResultsCount: 2,
+	}
+	res, err := w.Search(ctx, params)
+	fmt.Println(res)
+	return err
+}
+
 func Main(ctx context.Context, observationCtx *observation.Context, ready service.ReadyFunc, config *Config) error {
 	logger := observationCtx.Logger
 
diff --git a/enterprise/cmd/embeddings/shared/service.go b/enterprise/cmd/embeddings/shared/service.go
index abed83d941..d555d4c470 100644
--- a/enterprise/cmd/embeddings/shared/service.go
+++ b/enterprise/cmd/embeddings/shared/service.go
@@ -20,7 +20,7 @@ func (svc) Configure() (env.Config, []debugserver.Endpoint) {
 }
 
 func (svc) Start(ctx context.Context, observationCtx *observation.Context, ready service.ReadyFunc, config env.Config) error {
-	return Main(ctx, observationCtx, ready, config.(*Config))
+	return Main2(ctx, observationCtx, ready, config.(*Config))
 }
 
 var Service service.Service = svc{}
```

### CODE REVIEW

File: enterprise/cmd/embeddings/shared/weaviate.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	getQueryEmbedding getQueryEmbeddingFn,
+	url *url.URL,
+) *weaviateClient {
+	if url == nil {
+		return &weaviateClient{
+			clientErr: errors.New("weaviate client is not configured"),
+		}
+	}
+
+	client, err := weaviate.NewClient(weaviate.Config{
```
Comment (stefanhengl) [inline enterprise/cmd/embeddings/shared/weaviate.go:42]: Note to our future selves: During my testing I believe I found that NewClient doesn't return an error even if it couldn't connect to a weaviate instance.

[In the docs](https://weaviate.io/developers/weaviate/client-libraries/go#installation-and-setup) (which seem slightly outdated) , they call

```
   schema, err := client.Schema().Getter().Do(context.Background())
    if err != nil {
        panic(err)
    }
```

to verify that the client can connect. Maybe we should use that as client err?

File: enterprise/cmd/embeddings/shared/BUILD.bazel [hunk 2]
```diff
@@ -45,6 +47,10 @@ go_library(
         "//lib/errors",
         "@com_github_hashicorp_golang_lru_v2//:golang-lru",
         "@com_github_sourcegraph_log//:log",
+        "@com_github_weaviate_weaviate//entities/models:go_default_library",
```
Comment (stefanhengl) [inline enterprise/cmd/embeddings/shared/BUILD.bazel:50]: I assume the changes here are generated?

Comment (keegancsmith) [reply enterprise/cmd/embeddings/shared/BUILD.bazel:50]: yup. you need to run `bazel configure` these days when adding files or 3rd party deps to keep bazel ticking along.

File: enterprise/cmd/embeddings/shared/weaviate.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+func (w *weaviateClient) Use(ctx context.Context) bool {
+	return featureflag.FromContext(ctx).GetBoolOr("search-weaviate", false)
+}
+
+func (w *weaviateClient) Search(ctx context.Context, params embeddings.EmbeddingsSearchParameters) (*embeddings.EmbeddingSearchResults, error) {
+	if w.clientErr != nil {
+		return nil, w.clientErr
+	}
+
+	embeddedQuery, err := w.getQueryEmbedding(params.Query)
```
Comment (camdencheek) [inline enterprise/cmd/embeddings/shared/weaviate.go]: I think this will need updated after [this](https://github.com/sourcegraph/sourcegraph/pull/50846)

**Review by camdencheek — APPROVED** [review]

File: enterprise/cmd/embeddings/shared/weaviate.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+		logger:            logger.Scoped("weaviate", "client for weaviate embedding index"),
+		readFile:          readFile,
+		getQueryEmbedding: getQueryEmbedding,
+		client:            client,
+		clientErr:         err,
+	}
+}
+
+func (w *weaviateClient) Use(ctx context.Context) bool {
+	return featureflag.FromContext(ctx).GetBoolOr("search-weaviate", false)
```
[the remark this answers is not in the document]
Comment (keegancsmith) [reply enterprise/cmd/embeddings/shared/weaviate.go:57]: > (Have we checked this actually works?)

I feel seen. One of the reasons I haven't landed this yet is that I haven't yet exercised that code path for real :D Will do before landing.

File: enterprise/cmd/embeddings/shared/weaviate.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	queryBuilder := func(typ string, limit int) *graphql.GetBuilder {
+		return graphql.NewQueryClassBuilder(typ).
+			WithNearVector((&graphql.NearVectorArgumentBuilder{}).
+				WithVector(embeddedQuery)).
+			WithWhere(filters.Where().
+				WithOperator(filters.Equal).
+				WithPath([]string{"repo"}).
+				WithValueInt(int64(params.RepoID))).
```
[the remark this answers is not in the document]
Comment (keegancsmith) [reply enterprise/cmd/embeddings/shared/weaviate.go:77]: yup properties are indexed by default with a classic inverted index. https://weaviate.io/developers/weaviate/configuration/indexes#configure-the-inverted-index

**Comment by keegancsmith** [comment]
The current state is we may adjust how we insert data into weaviate. So this PR will need updating. However, I think those changes will be tiny so to reduce review toil my plan is the following:
- ensure feature flags work in embeddings
- land this PR as is
- follow-up with a tiny PR based on changes we make to the data model in weaviate.

**Review by mrnugget — APPROVED** [review]

**Pull request merged**